File: src/GUI/objects/ItemInfo.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package GUI.objects;

import static AudioPlayer.tagging.Cover.Cover.CoverSource.ANY;
import AudioPlayer.tagging.Metadata;
import java.io.IOException;
import java.net.URL;
import java.util.Objects;
import java.util.ResourceBundle;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.scene.control.Label;
import javafx.scene.layout.AnchorPane;

/**
 * FXML Controller class
 *
 * @author Plutonium_
 */
public class ItemInfo extends AnchorPane implements Initializable {
    
    @FXML private Label typeL;
    @FXML private Label indexL;
    @FXML private Label songL;
    @FXML private Label artistL;
    @FXML private Label albumL;
    @FXML private AnchorPane infoContainer;
    @FXML private AnchorPane coverContainer;
    private Thumbnail thumb;

    public ItemInfo() {
        this(true);
    }
    
    public ItemInfo(boolean include_cover) {
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(ItemInfo.class.getResource("ItemInfo.fxml"));
            fxmlLoader.setRoot(this);
            fxmlLoader.setController(this);
            fxmlLoader.load();
            
            if(include_cover) {
                // create
                thumb = new Thumbnail(coverContainer.getPrefHeight());
                coverContainer.getChildren().add(thumb.getPane());
            } else {
                // remove cover
                getChildren().remove(coverContainer);
                AnchorPane.setLeftAnchor(infoContainer, AnchorPane.getRightAnchor(infoContainer));
                coverContainer = null;
            }
            
        } catch (IOException e) {
            throw new RuntimeException("Could not load ItemInfo.fxml", e);
        }
    }
    
    /** {@inheritDoc} */
    @Override
    public void initialize(URL url, ResourceBundle rb) {}
    
    /**
     * Displays metadata information and title.
     * @param title
     * @param m 
     */
    public void setData(String title, Metadata m) {
        Objects.requireNonNull(m);
        typeL.setText(title);
        if (thumb!=null) thumb.loadImage(m.getCover(ANY));
        indexL.setText(m.getPlaylistIndexInfo());
        songL.setText(m.getTitle());
        artistL.setText(m.getArtist());
        albumL.setText(m.getAlbum());
    }
    
}

File: src/sp/it/pl/gui/infonode/ItemInfo.kt
package sp.it.pl.gui.infonode

import javafx.fxml.FXML
import javafx.geometry.Pos
import javafx.scene.control.Label
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.HBox
import javafx.scene.layout.Pane
import sp.it.pl.audio.Item
import sp.it.pl.audio.tagging.Metadata
import sp.it.pl.gui.objects.image.Thumbnail
import sp.it.pl.gui.objects.image.cover.Cover.CoverSource.ANY
import sp.it.pl.gui.objects.rating.Rating
import sp.it.pl.layout.widget.feature.SongReader
import sp.it.pl.util.async.FX
import sp.it.pl.util.async.future.Fut
import sp.it.pl.util.graphics.fxml.ConventionFxmlLoader
import sp.it.pl.util.graphics.setAnchors
import sp.it.pl.util.identityHashCode
import java.util.function.Consumer

/** Basic display for song information. */
class ItemInfo @JvmOverloads constructor(showCover: Boolean = true): HBox(), SongReader {

    @FXML private lateinit var indexL: Label
    @FXML private lateinit var songL: Label
    @FXML private lateinit var artistL: Label
    @FXML private lateinit var albumL: Label
    @FXML private var coverContainer: AnchorPane? = null
    private val rating = Rating()
    private val thumb: Thumbnail?
    private var dataId = null.identityHashCode()

    init {
        ConventionFxmlLoader(ItemInfo::class.java, this).loadNoEx<Any>()

        rating.alignment.value = Pos.CENTER_LEFT
        (albumL.parent as Pane).children.apply { add(indexOf(albumL), rating) }

        if (showCover) {
            thumb = Thumbnail()
            thumb.borderVisible = true
            coverContainer!!.children += thumb.pane
            thumb.pane.setAnchors(0.0)
        } else {
            thumb = null
            children -= coverContainer
            coverContainer = null
        }
    }

    override fun read(items: List<Item>) = read(items.firstOrNull() ?: Metadata.EMPTY)

    override fun read(m: Item) = setValue(m.toMeta())

    /** Displays metadata information. */
    fun setValue(m: Metadata) {
        dataId = m.identityHashCode()
        thumb?.loadCoverOf(m)
        indexL.text = m.getPlaylistIndexInfo().toString()
        songL.text = m.getTitle() ?: m.getFilename()
        artistL.text = m.getArtist()
        rating.rating.value = m.getRatingPercent()
        albumL.text = m.getAlbum()
    }

    private fun Thumbnail.loadCoverOf(data: Metadata) {
        val id = data.identityHashCode()
        Fut.futWith { data.getCover(ANY).image }
           .use(FX, Consumer { if (dataId==id) loadImage(it) })
    }

}